value/valuereflect_test.go

```diff
@@ -0,0 +1,520 @@
+			expectedMap:          map[string]interface{}{"int": int64(10), "S": "string"},
+			expectedUnstructured: map[string]interface{}{"int": int64(10), "S": "string"},
+		},
+		{
+			name:                 "omit",
+			val:                  testOmitStruct{I: 10, S: "string"},
+			expectedMap:          map[string]interface{}{"S": "string"},
+			expectedUnstructured: map[string]interface{}{"S": "string"},
+		},
+		{
+			name:                 "inline",
+			val:                  &testInlineStruct{Inline: T{I: 10}, S: "string"},
+			expectedMap:          map[string]interface{}{"int": int64(10), "S": "string"},
+			expectedUnstructured: map[string]interface{}{"int": int64(10), "S": "string"},
+		},
+		{
+			name:                 "omitempty",
+			val:                  testOmitemptyStruct{Noomit: nil, Omit: nil},
+			expectedMap:          map[string]interface{}{"noomit": (*string)(nil)},
+			expectedUnstructured: map[string]interface{}{"noomit": nil},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			rv := MustReflect(tc.val)
+			if !rv.IsMap() {
+				t.Error("expected IsMap to be true")
+			}
+			m := rv.AsMap()
+			if m.Length() != len(tc.expectedMap) {
+				t.Errorf("expected map to be of length %d but got %d", len(tc.expectedMap), m.Length())
+			}
+			iterateResult := map[string]interface{}{}
+			m.Iterate(func(s string, value Value) bool {
+				iterateResult[s] = value.(*valueReflect).Value.Interface()
+				return true
+			})
+			if !reflect.DeepEqual(iterateResult, tc.expectedMap) {
+				t.Errorf("expected iterate to produce %#v but got %#v", tc.expectedMap, iterateResult)
+			}
+
+			unstructured := rv.Unstructured()
+			if !reflect.DeepEqual(unstructured, tc.expectedUnstructured) {
+				t.Errorf("expected iterate to produce %#v but got %#v", tc.expectedUnstructured, unstructured)
+			}
+		})
+	}
+}
+
+type testMutateStruct struct {
+	I1 int64  `json:"key1,omitempty"`
+	S1 string `json:"key2,omitempty"`
+	S2 string `json:"key3,omitempty"`
+	S3 string `json:"key4,omitempty"`
+}
+
+func TestReflectStructMutate(t *testing.T) {
+	rv := MustReflect(&testMutateStruct{I1: 1, S1: "string1"})
+	if !rv.IsMap() {
+		t.Error("expected IsMap to be true")
+	}
+	m := rv.AsMap()
+	atKey1, ok := m.Get("key1")
+	if !ok {
+		t.Fatalf("expected map.Get(key1) to be 1 but got !ok")
+	}
+	if atKey1.AsInt() != 1 {
+		t.Fatalf("expected map.Get(key1) to be 1 but got: %v", atKey1)
+	}
+	m.Set("key1", NewValueInterface(int64(2)))
+	m.Delete("key2")
+	m.Delete("key3")
+	m.Set("key4", NewValueInterface("string4"))
+
+	expectedMap := map[string]interface{}{"key1": int64(2), "key4": "string4"}
+	unstructured := rv.Unstructured()
+	if !reflect.DeepEqual(unstructured, expectedMap) {
+		t.Errorf("expected %v but got: %v", expectedMap, unstructured)
+	}
+}
+
+// TestReflectMutateNestedStruct ensures a structs nested within various typed can be modified.
+func TestReflectMutateNestedStruct(t *testing.T) {
```

Comment:
That looks pretty exhaustive, thank you for writing this test. It's probably not needed, but I'm curious if we could write this test slightly more generically ... It doesn't have to be done now though.

Reply:
I went ahead and generalized it now.